xdggs/plotting.py

```diff
@@ -0,0 +1,53 @@
+import numpy as np
+
+
+def explore(
+    arr,
+    cell_boundaries="cell_boundaries",
+    cell_dim="cells",
+    cmap="viridis",
+    center=None,
+):
+    import geopandas as gpd
```

Comment:
In case you have interest, it would be possible to implement this conversion in a way that doesn't require GeoPandas (or pyarrow), by using raw numpy buffers and [arro3](https://github.com/kylebarron/arro3), a more minimal implementation of Arrow in Python that Lonboard adds as a dependency in the next version.

This would also be faster to render as it doesn't have to go through GeoPandas (and shapely geometries).

But it would be a little more complex code here. I can give you some pointers if you want.

Reply:
I'd definitely be interested.

Right now, what we're doing is compute cell vertices from the cell ids (which can have various forms for different DGGS, but most often are `int64` or `uint64` arrays), then construct them into `shapely` polygons. This is one of the things that take a longest other than converting to `geopandas` and constructing the `SolidPolygonLayer`, so if we could instead extend `deck.gl` to directly support them (like it supports H3), that would be best. However, I suppose that will either take a long time and/or require someone to actually do the work (and I'm not confident in my TypeScript / JavaScript skills so I won't be able to work on this myself anytime soon).

So as the next best step, should we construct `geoarrow` polygons from the cell vertex coordinates, possibly using `arro3`? If you have any examples I'd be happy to try figuring out how to do that.

Reply:
At the end of the day, we need to create vertex information in Lonboard/deck.gl for rendering. I believe that even deck.gl's H3 layer does not render the h3 cells on the GPU from an ID, but rather uses those IDs on the CPU in JS to create the vertex geometry.

For now, I think it's simplest to convert cell ids to Polygon geometries in Python, and then we can use the `SolidPolygonLayer`/`PolygonLayer` directly.

A GeoArrow Polygon array is just:

- a contiguous buffer of all the coordinates in all the polygons
- plus an offset array that points into the coordinates array and which says which coordinates are in each ring
- plus an offset array that points into the previous ring offsets array, and which says how many rings are in each geometry.

Here's a quick code example that might 

```py
import arro3.core
import h3.api.numpy_int as h3
import lonboard
import numpy as np
from arro3.core import Array, Field, fixed_size_list_array, list_array

cell = h3.geo_to_h3(-10, 15, 4)
disk = h3.k_ring(cell, 2)

# Overallocates in case there are any pentagons
coords_per_hexagon = 7
total_coords = len(disk) * coords_per_hexagon
# Allocate coordinate buffer
coords = np.zeros((total_coords, 2), dtype=np.float64)
# Allocate ring offsets. The length should be one greater than the number of rings
ring_offsets = np.zeros(len(disk) + 1, dtype=np.int32)
# In the case of h3 cells, we know that every cell will be a simple polygon with a
# single ring, so we can just use `arange`
geom_offsets = np.arange(len(disk) + 1, dtype=np.int32)

coord_offset = 0
for i, h in enumerate(disk):
    boundary = h3.h3_to_geo_boundary(h, geo_json=True)
    num_coords = len(boundary)
    coords[coord_offset : coord_offset + num_coords, :] = boundary
    coord_offset += num_coords
    ring_offsets[i + 1] = coord_offset

# shrink the coordinates array, in case there are any pentagons
last_coord = ring_offsets[-1]
coords = coords[:last_coord, :]


polygon_array = list_array(geom_offsets, list_array(ring_offsets, coords))
polygon_array_with_geo_meta = polygon_array.cast(
    polygon_array.field.with_metadata({"ARROW:extension:name": "geoarrow.polygon"})
)
lonboard.viz(polygon_array_with_geo_meta)
```

(Note that this simple implementation probably breaks over the antimeridian.)

h3-py is not itself vectorized, so converting from h3 cell ids to polygon vertices is quite slow. In the longer term, I'd like to update https://github.com/nmandery/h3ronpy to have a fast way to convert from h3 cell ids to GeoArrow polygon array (or MultiPolygonArray in the case of cells that cross the antimeridian)

Reply:
That looks great @kylebarron !

> I believe that even deck.gl's H3 layer does not render the h3 cells on the GPU from an ID, but rather uses those IDs on the CPU in JS to create the vertex geometry.

Interesting. Do you know if deck.gl has some clever optimizations for plotting H3/S2 like, e.g., compute on the fly the cell polygon geometries and/or aggregating data (parent cells), depending on the current zoom level / display extent? This might still have some benefits for plotting very large datasets (common use case) interactively. Or maybe Lonboard allows doing such optimizations on the server side (widget callback)?

>  I just implemented support for using numpy-backed memory for Rust Arrow arrays

I guess this could really help better integration between arrow and xarray in general? 👀

Reply:
> Interesting. Do you know if deck.gl has some clever optimizations for plotting H3/S2 

I think mostly https://deck.gl/docs/api-reference/geo-layers/h3-hexagon-layer#highprecision

> compute on the fly the cell polygon geometries 

It computes vertices here: https://github.com/visgl/deck.gl/blob/da0c1f9e3a23c43d87b062c41b5d5f908a4f6aef/modules/geo-layers/src/h3-layers/h3-hexagon-layer.ts#L182-L222

> aggregating data (parent cells), depending on the current zoom level / display extent? 

I think this is expected to happen server side if the user desires. At least the `H3HexagonLayer` won't do this automatically.

> Or maybe Lonboard allows doing such optimizations on the server side

Potentially Lonboard could do a compaction step (through h3ronpy) as an optional preprocess step.

> > I just implemented support for using numpy-backed memory for Rust Arrow arrays
> 
> I guess this could really help better integration between arrow and xarray in general?

I'm not sure whether `pyarrow` will reuse numpy memory when you pass in a numpy array as input. But at least for Arrow Rust it should be possible, even with multidimensional array input, as long as the Numpy data is numeric and C contiguous. Multidimensional arrays get converted to nested arrow FixedSizeList arrays, and there should be no overhead there.

Reply:
> `h3ronpy` has `h3ronpy.arrow.vector.cells_to_wkb_polygons` which would require you to decode WKB, so not sure if that's better.

No, that's not really better, at least as of today, because for WKB input we parse to shapely and then use `shapely.to_ragged_array` to convert to Arrow. In the future, we'll use Python bindings to geoarrow-rs to parse WKB directly to GeoArrow, but I need to test the 3D WKB parsing a bit more before putting it in prod in Lonboard.

> which takes ~19 s to complete, plus about 10s to render. The version that converts to geopandas takes ~54s plus another ~10s to render, so this is a huge improvement.

Can you check which part of the 19s is in which part? I'm curious how much of that is the `cell_ids2vertices`. Every other part of that code snippet should be very fast.

How many cell ids is this?

> ```py
> # fix the dateline / prime meridian issue
> lon_ = boundaries[..., 0]
> to_fix = abs(np.max(lon_, axis=-1) - np.min(lon_, axis=-1)) > 300
> fixed_lon = lon_[to_fix, :] + 360
> boundaries[to_fix, :, 0] = fixed_lon
> ```

I'm not sure this will work. I don't know if deck.gl will render coordinates outside of `[-180, 180]`. You might need to split along the antimeridian into a MultiPolygon.

Reply:
> How many cell ids is this?

`12 * 4**10` cell ids (roughly 12.5M)

> I'm not sure this will work. I don't know if deck.gl will render coordinates outside of [-180, 180].

`deck.gl` does render coordinates outside of `[-180, 180]` but has the snapping issue we talked about in #11.

And it's true the dateline issue is almost but not quite resolved: the cells crossing over the dateline now go from slightly more than 180 to slightly less than 480, so I'm missing a modulo somewhere.

> Can you check which part of the 19s is in which part?

I'll start profiling then.

Reply:
> > I'm not sure this will work. I don't know if deck.gl will render coordinates outside of [-180, 180].
> 
> `deck.gl` does render coordinates outside of `[-180, 180]` but has the snapping issue we talked about in #11.

Oh, that makes a lot more sense then. I don't think deck.gl _intentionally_ supports coordinates outside of `[-180, 180]`, and so you're probably getting an edge case from trying to render a different coordinate range.

> > Can you check which part of the 19s is in which part?
> 
> I'll start profiling then.

I took a quick look at this myself and it looked like there was a surprising amount of overhead in just calling `SolidPolygonLayer`. I should spend some time benchmarking that too

Reply:
yeah, I just came to the same conclusion. Here's the output of `pyinstrument`:
```python
19.600 <module>  lonboard_example.py:1
├─ 11.166 SolidPolygonLayer.__init__  lonboard/_layer.py:1498
│     [26 frames hidden]  lonboard, ipywidgets, pyarrow, <__arr...
│        8.384 ParquetWriter.write_table  pyarrow/parquet/core.py:1084
├─ 6.762 cell_ids2cell_boundaries_geoarrow  lonboard_example.py:36
│  ├─ 6.479 cell_ids2vertices  lonboard_example.py:18
│  │  ├─ 3.559 [self]  lonboard_example.py
│  │  ├─ 1.913 vec2ang  healpy/pixelfunc.py:730
│  │  │     [7 frames hidden]  healpy, <__array_function__ internals...
│  │  ├─ 0.411 amin  <__array_function__ internals>:177
│  │  │     [3 frames hidden]  numpy, <built-in>
│  │  └─ 0.408 amax  <__array_function__ internals>:177
│  │        [3 frames hidden]  numpy, <built-in>
│  └─ 0.253 concatenate  <__array_function__ internals>:177
│        [1 frames hidden]  <built-in>
├─ 0.845 <module>  healpy/__init__.py:1
│     [4 frames hidden]  healpy, astropy
└─ 0.585 <module>  lonboard/__init__.py:1
      [5 frames hidden]  lonboard, ipywidgets, IPython, pandas
```

Reply:
That's cool! I should figure out how to use `pyinstrument`. Was it a one-line thing to create that?

```
├─ 11.166 SolidPolygonLayer.__init__  lonboard/_layer.py:1498
│     [26 frames hidden]  lonboard, ipywidgets, pyarrow, <__arr...
│        8.384 ParquetWriter.write_table  pyarrow/parquet/core.py:1084
```

1. That's a lot of time just to write Parquet. I wouldn't have expected it to be that slow. For a long time I've thought about having a toggle for the user to switch off parquet encoding (e.g. if the Python environment is on the same machine as the browser). We go through Parquet currently because it creates a much smaller download in the case of remote Python environments.
2. I'm curious what those other frames are. It still shouldn't be 3 seconds of overhead

Reply:
> Was it a one-line thing to create that?

it has different modes you can run it (see the [user guide](https://pyinstrument.readthedocs.io/en/latest/guide.html)). I put the code above in a python script, then ran it with `python -m pyinstrument lonboard_example.py`, but you could also use the context manager:
```python
with pyinstrument.profile():
    ...
```
or the jupyter / ipython cell magic `%%pyinstrument` (load first with `%load_ext pyinstrument`)

Reply:
> I'm curious what those other frames are. It still shouldn't be 3 seconds of overhead

the ones beneath `SolidPolygonLayer`? Here you go:
```
─ 11.018 SolidPolygonLayer.__init__  lonboard/_layer.py:1498
│  └─ 11.018 SolidPolygonLayer.__init__  lonboard/_layer.py:266
│     ├─ 9.058 SolidPolygonLayer.__init__  lonboard/_layer.py:80
│     │  └─ 9.058 SolidPolygonLayer.__init__  lonboard/_base.py:16
│     │     └─ 9.058 SolidPolygonLayer.__init__  ipywidgets/widgets/widget.py:500
│     │           [1 frames hidden]  ipywidgets
│     │              8.991 SolidPolygonLayer.get_state  ipywidgets/widgets/widget.py:589
│     │              ├─ 8.330 serialize_table  lonboard/_serialization.py:72
│     │              │  └─ 8.330 serialize_table_to_parquet  lonboard/_serialization.py:25
│     │              │     └─ 8.321 ParquetWriter.write_batch  pyarrow/parquet/core.py:1068
│     │              │           [0 frames hidden]  
│     │              │              8.321 ParquetWriter.write_table  pyarrow/parquet/core.py:1084
│     │              └─ 0.661 serialize_accessor  lonboard/_serialization.py:58
│     │                 └─ 0.661 serialize_pyarrow_column  lonboard/_serialization.py:52
│     │                    └─ 0.661 serialize_table_to_parquet  lonboard/_serialization.py:25
│     │                       └─ 0.641 ParquetWriter.write_batch  pyarrow/parquet/core.py:1068
│     │                             [1 frames hidden]  pyarrow
│     └─ 1.958 default_geoarrow_viewport  lonboard/_layer.py:215
│        └─ 1.865 total_bounds  lonboard/_geoarrow/ops/bbox.py:36
│           └─ 1.865 _total_bounds_nest_2  lonboard/_geoarrow/ops/bbox.py:80
│              └─ 1.865 _coords_bbox  lonboard/_geoarrow/ops/bbox.py:55
│                 ├─ 0.939 amax  <__array_function__ internals>:177
│                 │     [3 frames hidden]  numpy, <built-in>
│                 └─ 0.926 amin  <__array_function__ internals>:177
│                       [3 frames hidden]  numpy, <built-in>
```

Reply:
FWIW the next version of `geoarrow.rust.core` will have constructor functions: https://github.com/geoarrow/geoarrow-rs/pull/810. And so you can pass the arrays directly into `geoarrow.rust.core.polygons`. Using those functions would ensure that the arrays generated are _fully_ spec-compatible. (As it is, you're missing the correct field names for nested arrays and the correct nullability flags for nested arrays. Lonboard allows that input but it's not 100% to the geoarrow spec, especially if we enforce it in issues like https://github.com/geoarrow/geoarrow/issues/64)